### ISSUE TITLE
About using vLLM for generation

### ISSUE THREAD

**Issue by LSC527** [OpenRLHF/OpenRLHF#188]
I have some thoughts about using vLLM for generation. Feel free to correct me if I were wrong.
1. Batching
It seems that prompts are still passing to vllm engines in micro rollout batches during `make_experience`.
However, passing all prompts to vllm engines all at once is very likely to imporve generation throughput.
2. Placement 
The device placement of vLLM engines seems quite random. For example,  this is what happens when running `examples/scripts/train_ppo_llama_ray_70b.sh`: 
run 1, matser node:
```
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    780546      C   ray::CriticModelRayActor         2806MiB |
|    1   N/A  N/A    780769      C   ray::CriticModelRayActor         2970MiB |
|    2   N/A  N/A    780770      C   ray::CriticModelRayActor         2990MiB |
|    3   N/A  N/A    780771      C   ray::CriticModelRayActor         2798MiB |
|    4   N/A  N/A    781017      C   ray::RewardModelRayActor         2530MiB |
|    5   N/A  N/A    781612      C   ray::RewardModelRayActor         2526MiB |
|    6   N/A  N/A    787426      C   ray::RayWorkerVllm              74344MiB |
|    7   N/A  N/A    787427      C   ray::RayWorkerVllm              74264MiB |
+-----------------------------------------------------------------------------+
```
run 2, matser node:
```
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    531162      C   ray::ActorModelRayActor.fit      2822MiB |
|    1   N/A  N/A    531384      C   ray::ActorModelRayActor.fit      3014MiB |
|    2   N/A  N/A    531385      C   ray::ActorModelRayActor.fit      3014MiB |
|    3   N/A  N/A    531386      C   ray::ActorModelRayActor.fit      2822MiB |
|    4   N/A  N/A    531387      C   ray::CriticModelRayActor         2824MiB |
|    5   N/A  N/A    532043      C   ray::CriticModelRayActor         3016MiB |
|    6   N/A  N/A    532044      C   ray::CriticModelRayActor         3016MiB |
|    7   N/A  N/A    532045      C   ray::CriticModelRayActor         2824MiB |
+-----------------------------------------------------------------------------+
```
To reduce communication overhead of broadcasting parameters from actor models to vllm engines, vllm engines and actor models are expected to be placed as close as possible. (e.g. on the same node)
The ideal device placement of this training task would be:
```
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    531162      C   ray::ActorModelRayActor.fit      2822MiB |
|    1   N/A  N/A    531384      C   ray::ActorModelRayActor.fit      3014MiB |
|    2   N/A  N/A    531385      C   ray::ActorModelRayActor.fit      3014MiB |
|    3   N/A  N/A    531386      C   ray::ActorModelRayActor.fit      2822MiB |
|    4   N/A  N/A    787426      C   ray::RayWorkerVllm              74344MiB |
|    5   N/A  N/A    787427      C   ray::RayWorkerVllm              74264MiB |
|    6   N/A  N/A    787426      C   ray::RayWorkerVllm              74344MiB |
|    7   N/A  N/A    787427      C   ray::RayWorkerVllm              74264MiB |
+-----------------------------------------------------------------------------+
```

**Comment by wuxibin89** [comment]
> 1. It seems that prompts are still passing to vllm engines in micro rollout batches during make_experience.
However, passing all prompts to vllm engines all at once is very likely to imporve generation throughput.

In current implementation, each actor model gets a micro rollout batch and independently sends to vllm engine for text generation. After generation, the micro batch is send to actor/critic/reward/reference to generate action_logits/ref_logits/values/rewards concurrently. 
If sending all prompts to vllm engines at once, then the 4 models are idle before generation done. A better approach is making these two stages pipelining and fully utilize all GPUs.

**Comment by wuxibin89** [comment]
> To reduce communication overhead of broadcasting parameters from actor models to vllm engines, vllm engines and actor models are expected to be placed as close as possible. (e.g. on the same node)

Nice catch! I will work on it.

**Comment by LSC527** [comment]
> > 1. It seems that prompts are still passing to vllm engines in micro rollout batches during make_experience.
> >    However, passing all prompts to vllm engines all at once is very likely to imporve generation throughput.
> 
> In current implementation, each actor model gets a micro rollout batch and independently sends to vllm engine for text generation. After generation, the micro batch is send to actor/critic/reward/reference to generate action_logits/ref_logits/values/rewards concurrently. If sending all prompts to vllm engines at once, then the 4 models are idle before generation done. A better approach is making these two stages pipelining and fully utilize all GPUs. This a baseline implementation for now, we're very welcome for your contribution.

我意思是可以一次性把`rollout_batch_size/vllm_num_engines`个prompt输入到每个vllm engine，而不是`micro_rollout_batch_size/vllm_num_engines`个prompt输入到每个vllm engine。
`make_experience`和`ppo_train`之间应该不是异步的吧？这样的话一次性generate一整个rollout_batch应该也不会引入额外的idle?

**Comment by hijkzzz** [comment]
很有道理~充分利用 vllm 的动态batch
